Added general python diagnostic parts

schlunma opened the pull request:
These are the non-diagnostic-specific changes of PR 68 in the private repository (developed for #352). I added new general python diagnostic functions which can be tested by running the example python namelist/diagnostic.

schlunma commented:
Working on the Codacy issues.

bouweandela commented on esmvaltool/cmor/_fixes/CMIP5/MIROC_ESM.py:
```diff
@@ -101,9 +101,6 @@ def fix_metadata(self, cube):
             if time.units.origin == 'days since 0000-01-01 00:00:00':
                 time.units = cf_units.Unit(
                     'days since 1849-01-01 00:00:00', calendar=calendar)
```
@jvegasbsc Does this work as expected?

jvegasbsc replied:
We have not been able to test this with actual data. This was ported from version 1 and it looks strange because it is setting  1849 instead of 1850, making me think that maybe there was a different reason for this fix. I will prefer to keep it here until we find that is giving us problems

bouweandela commented on esmvaltool/diag_scripts/shared/plot/styles_python/import_ncl_style.py:
```diff
@@ -0,0 +1,143 @@
+                    else:
+                        infos.update({FILLING: 'none'})
+
+                    # Shape
```
Replace the huge if/else tree by a dictionary, e.g.:
```
shape = {
    0: 'x',
    1: '.',
    2: '+',
}
info = shape[info]
```

schlunma commented:
> Would it be possible to use yaml instead of the .ini format for the styles? That would be more consistent, because we use yaml everywhere.

Done.

> I'm not entirely convinced that the new example diagnostic is better or easier to understand. Maybe it would be better to have two copies of the diagnostic.py script, one simple as we had before, and one using the Models and Variables class, e.g. called diagnostic_object_based.py or something similar.

Done.

> Note that it is possible to chain diagnostic scripts from the recipe, but in that case only the first diagnostic script after the preprocessor will receive the configuration with a filled 'input_data' key, the next scripts will just receive a list of 'work_dir's of it's ancestor diagnostic scripts in `input_files`.

The classes can still be applied in this case, but no data is read. In this case, a warning is printed.

bouweandela commented on esmvaltool/diag_scripts/shared/__init__.py:
```diff
@@ -1,8 +1,62 @@
 """Code that is shared between multiple diagnostic scripts."""
 from . import plot
 from ._base import get_cfg, run_diagnostic
+from .python_diag import (
```
I think it would be cleaner if you could move all of these names into a `esmvaltool.diag_scripts.shared.names` module. Actually, I think it may be a bit overkill to do it like this, but if you prefer it this way I'm fine with including the names as constants, but please in a separate module.

bouweandela commented:
Thanks for making the requested changes. I have one more comment (above) and there is the comment by @jvegasbsc which still needs to be addressed: do you have a good reason for removing that fix? If so, maybe it is better to open a separate issue and pull request for it, because it will make it easier to merge this pull request because then we don't have to wait for it.

Please also make sure the unit tests pass.

One more minor thing: I realised that the commonly used term is object oriented instead of object based, so maybe that would be a slightly better name for this object oriented example diagnostic.

schlunma commented:
> I think it would be cleaner if you could move all of these names into a esmvaltool.diag_scripts.shared.names module. Actually, I think it may be a bit overkill to do it like this, but if you prefer it this way I'm fine with including the names as constants, but please in a separate module.

Constant names can now be accessed via the `esmvaltool.diag_scripts.shared.names` module.

> Thanks for making the requested changes. I have one more comment (above) and there is the comment by @jvegasbsc which still needs to be addressed: do you have a good reason for removing that fix? If so, maybe it is better to open a separate issue and pull request for it, because it will make it easier to merge this pull request because then we don't have to wait for it.

The changes were made by @jvegasbsc, the only reason was that without these changes the specific model could not be handled by the preprocessor. However, I reverted them and removed the model from the recipe. I will open an issue about the two models which fail in my recipe

> Please also make sure the unit tests pass.

Done (the reason for the failing tests was the modified fix file for MIROC-ESM.

> One more minor thing: I realised that the commonly used term is object oriented instead of object based, so maybe that would be a slightly better name for this object oriented example diagnostic.

Done.

bouweandela commented:
It looks like all public functionality of the `esmvaltool.diag_scripts.shared.python_diag` is available by importing from `esmvaltool.diag_scripts.shared`, so I think you can make it private by giving it a name starting with an underscore. Also, having the word python in there is probably not needed as the extension is already .py. Therefore I would suggest renaming esmvaltool/diag_scripts/shared/python_diag.py to esmvaltool/diag_scripts/_diag.py. After that, I think this is ready to be merged.
